keps/sig-storage/4958-csi-sidecars-all-in-one/README.md

```diff
@@ -0,0 +1,1214 @@
+    - Any new feature PRs are not allowed to be filed to the master branch or release-X branches(Controlled by the individual repo maintainer, categorize it and reject it if it's a feature)
+    - SIG Storage Developer file the feature PRs to AIO MonoRepo 
+    - Except for the serious bugfixes or CVE fixes PRs (only from individual repo maintainer) which can be merged in master and backported to the other release-X branches
+- Deprecated:
+    - Not maintaining this repository
+    - Eventually the image is going away for the individual repo is going away (although wouldn't possible unless we migrate ALL the sidecars)
+    - (future) archive it but not at the same time as the deprecation time, this is a terminal state so we can't undo it
+
+
+![state change](./aio11.png "overview of workflow definition")
+
+##### Migration Process
+
+![migration process](./aio10.png "")
+
+
+#### Risks And Mitigations
+
+- Breaking changes in one component forces the single release to be a breaking change
+
+- Vulnerability that might affects one component affects all other components
+
+see details in: https://docs.google.com/document/d/1SD4YRas_qXMP363L4j3WBTV_F9anq-5FM5gdGmJq7h0/edit?usp=sharing
+
+- Panic in one component restarts the sidecar
+
+For each sidecar define the where in the stack a panic should be caught to possibly restart the controller. 
+
+List of fixed issues related with panics:
+    - https://github.com/kubernetes-csi/external-provisioner/issues/839
+    - https://github.com/kubernetes-csi/external-provisioner/issues/582 
+    - https://github.com/kubernetes-csi/external-attacher/issues/502
+
+> panic like OOM doesn't count into this type(perhaps no good way to reduce the blast radius)
+
+- Keeping the monorepo and the existing sidecars repo up to date after the migration for X releases
+
+
+### MileStone
+
+---
+design phase
+#### Milestone (completed): 
+Develop a minimal proof of concept
+
+POC: https://github.com/mauriciopoppe/csi-sidecars-aio-poc
+
+#### Milestone-definition-of-the-development-workflow(completed)
+
+--- 
+prepare phase
+
+#### Milestone-sidecar-entry-codes-have-been-modified.
+
+
+#### Milestone-setup-a-repository-inside-kubernetes-csi
+
+
+#### Milestone-mirroring-of-nested-directories-to-repos-in-kubernetes-csi
+
+
+#### Milestone-Build-the-project-using-a-modified-copy-of-release-tools
+
+
+#### Milestone-set-up-new-test-infra-jobs-to-test-the-project-through-the-hostpath-CSI-Driver
+
+
+#### Milestone-migration-of-CSI-Drivers-to-the-new-model 
+
+
+---
+
+
+#### Milestone-all-six-sidecar-repo-had-been-integrated-into-monorepo
+
+Alpha phase
+
+#### Milestone-be-ready-to-accept-PR-from-community
+
+Beta phase
+
+#### Milestone-six-sidecars-working-through-CSI-hostpath
+
+Beta phase
+
+#### Milestone-three-cloud-vendors-start-using-the-monorepo-component
+
+GA phase
+
+#### Milestone-all-individual-repo-has-been-into-deprecated-state
+
+Standalone phase
+
+#### Milestone-merge-sidecar-informer-caches
```

Comment:
This is a nice improvement that shouldn't be part of the MVP yet. It could be its own KEP, it could happen after all of the CSI sidecars have been deprecated or migrated to the monorepo, it could happen at many different points in time. It should be listed here though for review as possible future improvements.